### PR TITLE
feat:(FEC-11397): add support for VTT thumbnails

### PR DESCRIPTION
**Pull request by JonathanTGold**
## Context:

At present state, the Kaltura player can handle two inputs for the timeline thumbnails:

1.  The thumbnails generated by Kaltura (when Kaltura holds the content and can generate the images).
2.   In-Stream DASH thumbnails

This feature is to support a third option: VTT Thumbnails.

The VTT thumbnails data, similar in essence to VTT (WebVTT) captions, is an external file containing image (or image sprite) urls and data about the VOD position to use those thumbnails.

## Requirements:

We need to parse the VTT file and display the images on the timeline.
There are a few variations for VTT files, we will look at two of them:

1.  One image file for each time period.
2.  An image sprite, with specific co-ordinations inside the VTT file - e.g. xywh=1280,0,128,72

We need to enable configuring such an external VTT files (similar to the way we allow external captions)
If such VTT file is configured, it should overrule the Kaltura’s thumbnail (if available)

solevs: FEC-11397

Design Document: https://kaltura.atlassian.net/wiki/spaces/FEC/pages/3141927050/Add+support+for+External+VTT+Thumbnails+-+Design+Document
### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [x] new unit / functional tests have been added (whenever applicable)
- [x] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [x] Docs have been updated

### CODE REVIEW

File: flow-typed/types/thumbnail-vtt-cue.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+declare type PKThumbnailVttCue = {
+  startTime: number,
+  endTime: number,
+  imgUrl: string,
+  coordinates: {x: number, y: number} | null,
+  size: {height: number, width: number}
```
Comment (JonathanTGold) [inline flow-typed/types/thumbnail-vtt-cue.js:6]: add null type

File: src/player.js [hunk 2]
```diff
@@ -40,6 +40,7 @@ import ImageTrack from './track/image-track';
 import {ThumbnailInfo} from './thumbnail/thumbnail-info';
 import {EngineDecoratorManager} from './engines/engine-decorator-manager';
 import {filterTracksByRestriction} from './utils/restrictions';
+import {ExternalThumbnailsHandler} from './track/external-thumbnails-handler';
```
Comment (JonathanTGold) [inline src/player.js:43]: move to thumbnails folder

File: src/player.js [hunk 3]
```diff
@@ -437,6 +444,7 @@ export default class Player extends FakeEventTarget {
     this._createPlayerContainer();
     this._appendDomElements();
     this._externalCaptionsHandler = new ExternalCaptionsHandler(this);
+    this._externalThumbnailsHandler = new ExternalThumbnailsHandler(this);
```
Comment (JonathanTGold) [inline src/player.js:447]: pass just the relevant configuration directly to the load() and not save state at all
and make sure reset - reset the entire state

File: src/player.js [hunk 4]
```diff
@@ -1426,7 +1435,9 @@ export default class Player extends FakeEventTarget {
    * @return {?ThumbnailInfo} - Thumbnail info
    */
   getThumbnail(time: number): ?ThumbnailInfo {
-    if (this._engine) {
+    if (this._externalThumbnailsHandler.isUsingVttThumbnails()) {
```
Comment (JonathanTGold) [inline src/player.js:1438]: it overnight dash thumbnails - confluence

File: src/track/external-thumbnails-handler.js [hunk 5]
```diff
@@ -0,0 +1,263 @@
+import Player from '../player';
+import EventManager from '../event/event-manager';
+import FakeEventTarget from '../event/fake-event-target';
+import getLogger from '../utils/logger';
+import {Parser, StringDecoder} from './text-track-display';
+import * as Utils from '../utils/util';
+import {ThumbnailInfo} from '../thumbnail/thumbnail-info';
+import {Cue} from './vtt-cue';
+
+const DEFAULT_THUMBNAIL_SIZE: {width: number, height: number} = {
```
Comment (JonathanTGold) [inline src/track/external-thumbnails-handler.js:10]: remove - default is natural height and width

File: src/track/external-thumbnails-handler.js [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+  width: 164,
+};
+
+const VTT_WITH_IMG_SIZE: RegExp = /#xy=/i;
+const VTT_WITH_IMG_SIZE_AND_COORDS: RegExp = /#xywh=/i;
+
+class ExternalThumbnailsHandler extends FakeEventTarget {
+  /**
+   * constructor
+   * @param {Player} player - the player instance.
```
Comment (JonathanTGold) [inline src/track/external-thumbnails-handler.js:20]: remove

File: src/track/external-thumbnails-handler.js [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+   * @private
+   */
+  _cues: Array<PKThumbnailVttCue> = [];
+
+  /**
+   * base url of the images or the sprite img
+   * @type {string}
+   * @private
+   */
+  _imgBaseUrl: string;
```
Comment (JonathanTGold) [inline src/track/external-thumbnails-handler.js:62]: redundent

File: src/track/external-thumbnails-handler.js [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+  /**
+   * start the loading and parsing process of the vtt thumbnails file.
+   * @returns {void}
+   * @public
+   */
+  async load(): Promise<void> {
+    const thumbnailsConfig: PKExternalThumbnailsConfig = this._player.sources.thumbnails;
+    if (!thumbnailsConfig) {
+      return [];
```
Comment (JonathanTGold) [inline src/track/external-thumbnails-handler.js:79]: return

File: src/track/external-thumbnails-handler.js [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+  }
+
+  /**
+   * Make a request to download the vtt file.
+   * @param {PKExternalThumbnailsConfig} thumbnailsConfig - the thumbnails config object.
+   * @returns {Promise<string>} - resolves with the vtt string.
+   * @private
+   */
+  async _downloadVttFile(thumbnailsConfig: PKExternalThumbnailsConfig): Promise<string> {
+    return await Utils.Http.execute(thumbnailsConfig.vttUrl, {}, 'GET');
```
Comment (OrenMe) [inline src/track/external-thumbnails-handler.js:130]: add debug printout in case download fails

File: src/track/external-thumbnails-handler.js [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+  /**
+   * download and parse the vtt file
+   * @param {PKExternalThumbnailsConfig} thumbnailsConfig - the thumbnails user config
+   * @returns {Promise<void>} - resolve when the loading and parsing process is complete
+   * @private
+   */
+  async _downloadAndParseCues(thumbnailsConfig: PKExternalThumbnailsConfig): Promise<void> {
+    const VttStr: string = await this._downloadVttFile(thumbnailsConfig);
+    const cuesArray: Array<Cue> = await this._processVtt(VttStr);
```
Comment (OrenMe) [inline src/track/external-thumbnails-handler.js:119]: what happens if no file returned and string is empty?

File: src/track/external-thumbnails-handler.js [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+    return VTT_WITH_IMG_SIZE.test(cues[0].text) || VTT_WITH_IMG_SIZE_AND_COORDS.test(cues[0].text) || false;
+  }
+
+  /**
+   * format vtt cue into thumbnail cue - by extracting the img options metadata.
+   * @param {Cue} cue - a parsed VTTCue in the vtt text cue format
+   * @returns {PKThumbnailVttCue} - cue object contains the img metadata.
+   * @private
+   */
+  _extractCueMetadata(cue: Cue): PKThumbnailVttCue {
```
Comment (JonathanTGold) [inline src/track/external-thumbnails-handler.js:202]: no support for #wh vtt

File: src/player.js (hunk 4 above)
Comment (OrenMe) [reply src/player.js:1438]: overnight?

File: src/track/external-thumbnails-handler.js (hunk 8 above)
Comment (OrenMe) [reply src/track/external-thumbnails-handler.js:79]: the method signature returns a promise

**Review by OrenMe — CHANGES_REQUESTED** [review]

File: src/player.js (hunk 4 above)
Comment (JonathanTGold) [reply src/player.js:1438]: It also works during the day of course ... the comment was fixed

File: src/track/external-thumbnails-handler.js (hunk 8 above)
Comment (JonathanTGold) [reply src/track/external-thumbnails-handler.js:79]: and it dose return promise cause it async function

File: src/track/external-thumbnails-handler.js (hunk 9 above)
Comment (JonathanTGold) [reply src/track/external-thumbnails-handler.js:130]: added error handling which print the error

File: src/error/code.js [hunk 12]
```diff
@@ -98,6 +98,10 @@ const Code: CodeType = {
    * https://github.com/kaltura/playkit-js/blob/master/docs/configuration.md#configsourcescaptions
    */
   UNKNOWN_LANGUAGE: 2012,
+  /**
+   * invalid thumbnail vtt file format or invalid utl option.
```
Comment (yairans) [inline src/error/code.js:102]:
```suggestion
   * invalid thumbnail vtt file format or invalid url option.
```

File: src/thumbnail/external-thumbnails-handler.js [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+  }
+
+  /**
+   * find and return the img info for the requested time.
+   * @param {number} time - timing in th playback timeline in milliseconds.
+   * @returns {ThumbnailInfo | null} - the thumbnail img info for the requested time.
+   * @public
+   */
+  getThumbnail(time: number): ThumbnailInfo | null {
+    const cue: PKThumbnailVttCue = this._findCue(time);
```
Comment (yairans) [inline src/thumbnail/external-thumbnails-handler.js:70]:
```suggestion
    const cue: PKThumbnailVttCue = this._findCue(time, this._cues);
```

File: src/thumbnail/external-thumbnails-handler.js [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+   * @returns {ThumbnailInfo | null} - the thumbnail cue linked to that timing.
+   * @private
+   */
+  _findCue(time: number, cues: Array<PKThumbnailVttCue>): ThumbnailInfo | null {
+    let left = 0;
+    let right = cues.length - 1;
+    while (left <= right) {
+      const middle = Math.floor((left + right) / 2);
+      const potentialCueMatch: PKThumbnailVttCue = cues[middle];
+      if (time >= potentialCueMatch.startTime && time < potentialCueMatch.endTime) {
```
Comment (yairans) [inline src/thumbnail/external-thumbnails-handler.js:270]:
```suggestion
      if (time >= potentialCueMatch.startTime && time <= potentialCueMatch.endTime) {
```

File: src/thumbnail/external-thumbnails-handler.js [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+   * format vtt cue into thumbnail cue - by extracting the img options metadata.
+   * @param {VTTCue} vttCue - a parsed VTTCue in the vtt text cue format
+   * @param {PKExternalThumbnailsConfig} thumbnailsConfig - the thumbnails config
+   * @returns {PKThumbnailVttCue} - cue object contains the img metadata.
+   * @private
+   */
+  _extractCueMetadata(vttCue: VTTCue, thumbnailsConfig: PKExternalThumbnailsConfig): PKThumbnailVttCue {
+    const {startTime, endTime, text} = vttCue;
+    const {imgBaseUrl} = thumbnailsConfig;
+    const isVTTIncludesImgSize: boolean = VTT_INCLUDES_SIZE_ONLY.test(text);
```
Comment (yairans) [inline src/thumbnail/external-thumbnails-handler.js:221]:
```suggestion
    const isVTTIncludesImgSizeOnly: boolean = VTT_INCLUDES_SIZE_ONLY.test(text);
```

**Review by yairans — CHANGES_REQUESTED** [review]

**Review by yairans — DISMISSED** [review]

**Review by yairans — APPROVED** [review]

**Pull request merged**